core/src/main/scala/org/apache/spark/scheduler/cluster/cook/CoarseCookSchedulerBackend.scala

```diff
@@ -182,33 +197,55 @@ private[spark] class CoarseCookSchedulerBackend(
 
     val cmds = remoteConfFetch ++ environment ++ Seq(commandString, cleanup)
 
-    new Job.Builder()
+    val job = new Job.Builder()
       .setUUID(jobId)
       .setEnv(environment.asJava)
       .setUris(uris.asJava)
       .setCommand(cmds.mkString("; "))
-      .setMemory(sparkMesosScheduler.calculateTotalMemory(sc))
+      .setMemory(mesosBackend.calculateTotalMemory(sc))
       .setCpus(numCores)
       .setPriority(priority)
       .build()
+
+    executorsToJobs(slaveId.toString + "/" + taskId.toString) = jobId
+
+    job
   }
 
   def createRemainingJobs(): List[Job] = {
     def loop(coresRemaining: Int, jobs: List[Job]): List[Job] =
-      if (coresRemaining <= 0) jobs
-      else if (coresRemaining <= maxCoresPerJob) createJob(coresRemaining) :: jobs
-      else loop(coresRemaining - maxCoresPerJob, createJob(maxCoresPerJob) :: jobs)
-    loop(maxCores - totalCoresRequested, Nil).reverse
+      if (jobs.size == executorLimit) jobs
+      else
+        if (maxCoresPerJob <= 0) jobs
+        else if (coresRemaining <= maxCoresPerJob) createJob(coresRemaining) :: jobs
+        else loop(coresRemaining - maxCoresPerJob, createJob(maxCoresPerJob) :: jobs)
+
+    loop(maxCores - totalCoresRequested, List()).reverse
   }
 
   def requestRemainingCores() : Unit = {
     val jobs = createRemainingJobs()
-    totalCoresRequested += jobs.map(_.getCpus.toInt).sum
-    runningJobUUIDs = runningJobUUIDs ++ jobs.map(_.getUUID)
-    jobClient.submit(jobs.asJava, jobListener)
+
+    if (!jobs.isEmpty)
```

Comment:
What failed?

On Wed, May 25, 2016, 6:16 PM Piotr Solnica <notifications@github.com>
wrote:

> In
> core/src/main/scala/org/apache/spark/scheduler/cluster/cook/CoarseCookSchedulerBackend.scala
> <https://github.com/mojotech/spark/pull/1#discussion_r64663245>:
>
> >    }
> >
> >    def requestRemainingCores() : Unit = {
> >      val jobs = createRemainingJobs()
> > -    totalCoresRequested += jobs.map(_.getCpus.toInt).sum
> > -    runningJobUUIDs = runningJobUUIDs ++ jobs.map(_.getUUID)
> > -    jobClient.submit(jobs.asJava, jobListener)
> > +
> > +    if (!jobs.isEmpty)
>
> Test didn't pass without this conditional. I'll try to simplify it.
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/mojotech/spark/pull/1/files/d7d3110f1b513f75b9f86c9f785f443332e63c63#r64663245>
>